src/hooks/useValidateRepository.ts

```diff
@@ -0,0 +1,188 @@
+          setValidationState({
+            message: getErrorMessage(errorDetail, authSearchString),
+            uri: 'error',
+            token: getFieldValue('token') ? 'error' : '',
+            username: getFieldValue('username') ? 'error' : '',
+          });
+          return;
+        }
+        if (errorDetail.includes(uriSearchString)) {
+          setValidationState({
+            message: getErrorMessage(errorDetail, uriSearchString),
+            uri: 'error',
+          });
+          return;
+        }
+
+        if (errorDetail.includes(branchSearchString)) {
+          setValidationState({
+            message: getErrorMessage(errorDetail, branchSearchString),
+            uri: 'success',
+            token: getFieldValue('token') ? 'success' : '',
+            username: getFieldValue('username') ? 'success' : '',
+            branch: 'error',
+          });
+          return;
+        }
+
+        if (errorDetail.includes(pathSearchString)) {
+          setValidationState({
+            message: getErrorMessage(errorDetail, pathSearchString),
+            uri: 'success',
+            token: getFieldValue('token') ? 'success' : '',
+            username: getFieldValue('username') ? 'success' : '',
+            branch: getFieldValue('branch') ? 'success' : '',
+            path: 'error',
+          });
+          return;
+        }
+
+        setValidationState({
+          message: 'Unknown error',
+          uri: 'error',
+        });
+      } else {
+        setValidationState({
+          uri: 'success',
+          token: getFieldValue('token') ? 'success' : '',
+          username: getFieldValue('username') ? 'success' : '',
+          branch: getFieldValue('branch') ? 'success' : '',
+          path: getFieldValue('path') ? 'success' : '',
+        });
+      }
+    },
+    [setValidationState, getFieldValue]
+  );
+
+  const getValidationPayload = useCallback(() => {
+    return {
+      type: 'git',
+      ...fieldsNames.reduce((acc, curr) => {
+        if (curr === 'token' && getFieldValue(curr) === dummySecret && getFieldValue('tokenSecret')) {
+          return {
+            ...acc,
+            tokenSecret: getFieldValue('tokenSecret'),
+          };
+        }
+
+        if (curr === 'username' && getFieldValue(curr) === dummySecret && getFieldValue('usernameSecret')) {
+          return {
+            ...acc,
+            usernameSecret: getFieldValue('usernameSecret'),
+          };
+        }
+        return {...acc, [curr]: getFieldValue(curr)};
+      }, {}),
+    };
+  }, [getFieldValue]);
+
+  useDebounce(
+    () => {
+      if (!getFieldValue('uri')) {
+        return;
+      }
+
+      setValidationState({
+        uri: 'loading',
+        token: getFieldValue('token') ? 'loading' : '',
+        username: getFieldValue('username') ? 'loading' : '',
+      });
+
+      validateRepository(getValidationPayload()).then((res: any) => {
+        handleResponse(res);
```

Comment:
After the response is there, it should check if the values are the same. Otherwise, it will fall into a race condition:

- 0ms - User updates the data **(invalid)**
- 300ms - Started 1st validation request
- 400ms - User updates the data **(valid)**
- 700ms - Started 2nd validation request
- 900ms - 2nd validation request finished and the validation results were updated
- 1000ms - 1st validation request finished **and the validation results were incorrectly updated to invalid**

Reply:
Sorry, didn't quite get it. Which values and how it is supposed to be checked?

Reply:
The form values. To show different example:

* You set the branch to `branch-1`, API Call #1 is started
* You set the branch to `branch-2`, API Call #2 is started
* API Call #2 is finished - it will show (correctly) results for the latest `branch-2`
* API Call #1 is finished - it will replace the correct results, with results for the old `branch-1`

To avoid that, you need to either cancel the previous API call after changes or ignore it afterward (by i.e. checking if the values are still the same)

Reply:
Ok, I got it. Thanks